geos: set RPATH to /usr/local/lib/cockroach and update docker image

otan opened the pull request:
Set the RPATH to /usr/local/lib/cockroach (postgres does
/usr/local/lib/postgres) that becomes the default location to look for
CRDB binaries. This avoids needing to set the `LD_LIBRARY_PATH` to make
the shared libraries for GEOS to work when copy pasting our copies.

Also update the docker deploy container to initialize this directory and
install the shared images.

Release note (general change): Reconfigured the cockroach binary to set
the RPATH to /usr/local/lib/cockroach. The docker image that is
available with CockroachDB will also include the GEOS files necessary to
work with Geospatial types in CockroachDB.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50356)
<!-- Reviewable:end -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50356)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @otan, and @petermattis)

___
*[pkg/geo/geos/geos.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50356#-MA2Jl5U-SkgGjFs7GAx:-MA2Jl5U-SkgGjFs7GAy:b-v64h7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/84be660c32549b101396804699a8ddf0835d617b/pkg/geo/geos/geos.go#L29)):*
> ```Go
> 
> // #cgo CXXFLAGS: -std=c++14
> // #cgo !windows LDFLAGS: -ldl -Wl,-rpath,/usr/local/lib/cockroach
> ```

I'm not clear on why setting RPATH is necessary. RPATH controls where the dynamic loader (`ld.so`) will look for shared libraries. But since GEOS is being loaded explicitly I would think this is unnecessary. I'm probably missing something.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geos/geos.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50356#-MA2Jl5U-SkgGjFs7GAx:-MA2Rbgf42QS3pho7A7g:b-yobm78) ([raw file](https://github.com/cockroachdb/cockroach/blob/84be660c32549b101396804699a8ddf0835d617b/pkg/geo/geos/geos.go#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm not clear on why setting RPATH is necessary. RPATH controls where the dynamic loader (`ld.so`) will look for shared libraries. But since GEOS is being loaded explicitly I would think this is unnecessary. I'm probably missing something.
</blockquote></details>

the libgeos_c.so file doesn't know how to look for libgeos.so.3.8.1 without this when i tried it. the problem is that there's two files.


<!-- Sent from Reviewable.io -->

knz commented:
you could manually dlopen the dep before you dlopen the latter.

otan committed:
geos: set RPATH to /usr/local/lib/cockroach and update docker image

Set the RPATH to /usr/local/lib/cockroach (postgres does
/usr/local/lib/postgres) that becomes the default location to look for
CRDB binaries. This avoids needing to set the `LD_LIBRARY_PATH` to make
the shared libraries for GEOS to work when copy pasting our copies.

Also update the docker deploy container to initialize this directory and
install the shared images.

Release note (general change): Reconfigured the cockroach binary to set
the RPATH to /usr/local/lib/cockroach. The docker image that is
available with CockroachDB will also include the GEOS files necessary to
work with Geospatial types in CockroachDB.

otan commented:
> you could manually dlopen the dep before you dlopen the latter.
did this instead, moving to separate PR!
